File: 1-BigO/7-Factorial.js
// examples of a Factorial Time Complexity O(n!)

// O(n!) - Factorial Time Complexity

// example 1: generating permutations in O(n!)
// O(n!) + O(n!) = O(n!)
const example1 = (arr) => {
  if (!arr.length) return [[]];
  const first = arr.shift();
  const perms = example1(arr);
  const totalPerms = [];
  perms.forEach((perm) => {
    for (let i = 0; i <= perm.length; i++) {
      const newPerm = [...perm.slice(0, i), first, ...perm.slice(i)];
      totalPerms.push(newPerm);
    }
  });
  return totalPerms;
};

// example 2: generating subsets in O(n!)
// O(n!) + O(n!) = O(n!)
const example2 = (arr) => {
  if (!arr.length) return [[]];
  const first = arr.shift();
  const subsets = example2(arr);
  const newSubsets = subsets.map((subset) => [first, ...subset]);
  return [...subsets, ...newSubsets];
};

// example 3: brute-force for Traveling Salesman Problem in O(n!)
// O(n!) + O(n!) = O(n!)
function calculateTotalDistance(path, graph) {
  // Calculate total distance for the given path using graph information
}

function generateAllPermutations(cities) {
  // Generate all permutations of cities
}

function bruteForceTSP(cities, graph) {
  const permutations = generateAllPermutations(cities);
  let minDistance = Infinity;
  let optimalPath = [];

  for (const permutation of permutations) {
      const distance = calculateTotalDistance(permutation, graph);
      if (distance < minDistance) {
          minDistance = distance;
          optimalPath = permutation;
      }
  }

  return { path: optimalPath, distance: minDistance };
}

// example 4: Brute-Force Algorithm for Subset Sum Problem
function isSubsetSum(set, n, target) {
  if (target === 0) {
      return true;
  }
  if (n === 0 && target !== 0) {
      return false;
  }
  if (set[n - 1] > target) {
      return isSubsetSum(set, n - 1, target);
  }
  return isSubsetSum(set, n - 1, target) || isSubsetSum(set, n - 1, target - set[n - 1]);
}

console.log("example1([1, 2, 3]): ", example1([1, 2, 3]));
console.log("example2([1, 2, 3]): ", example2([1, 2, 3]));
console.log("bruteForceTSP([1, 2, 3]): ", bruteForceTSP([1, 2, 3]));

const cities = [1, 2, 3];
const graph = {
    1: { 2: 10, 3: 15 },
    2: { 1: 10, 3: 20 },
    3: { 1: 15, 2: 20 }
};
const result = bruteForceTSP(cities, graph);
console.log("Optimal TSP Path:", result.path);
console.log("Total Distance:", result.distance);

console.log("isSubsetSum([1, 2, 3]): ", isSubsetSum([1, 2, 3]));
const inputSet = [3, 34, 4, 12, 5, 2];
const targetSum = 9;
const isSubsetPossible = isSubsetSum(inputSet, inputSet.length, targetSum);
console.log(`Subset with sum ${targetSum} exists: ${isSubsetPossible}`);

File: 1-BigO/7-Factorial.py
# Example 1: Generating permutations
def example1(arr):
    if not arr:
        return [[]]
    first = arr.pop(0)
    perms = example1(arr)
    total_perms = []
    for perm in perms:
        for i in range(len(perm) + 1):
            new_perm = perm[:i] + [first] + perm[i:]
            total_perms.append(new_perm)
    return total_perms

# Example 2: Generating subsets
def example2(arr):
    if not arr:
        return [[]]
    first = arr.pop(0)
    subsets = example2(arr)
    new_subsets = [([first] + subset) for subset in subsets]
    return subsets + new_subsets

# Example 3: Brute-force for Traveling Salesman Problem
def calculate_total_distance(path, graph):
    # Calculate total distance for the given path using graph information
    pass

def generate_all_permutations(cities):
    # Generate all permutations of cities
    pass

def brute_force_tsp(cities, graph):
    permutations = generate_all_permutations(cities)
    min_distance = float('inf')
    optimal_path = []

    for permutation in permutations:
        distance = calculate_total_distance(permutation, graph)
        if distance < min_distance:
            min_distance = distance
            optimal_path = permutation

    return {'path': optimal_path, 'distance': min_distance}

# Example 4: Brute-force algorithm for Subset Sum Problem
def is_subset_sum(set, n, target):
    if target == 0:
        return True
    if n == 0 and target != 0:
        return False
    if set[n - 1] > target:
        return is_subset_sum(set, n - 1, target)
    return is_subset_sum(set, n - 1, target) or is_subset_sum(set, n - 1, target - set[n - 1])

print("example1([1, 2, 3]): ", example1([1, 2, 3]))
print("example2([1, 2, 3]): ", example2([1, 2, 3]))
print("brute_force_tsp([1, 2, 3]): ", brute_force_tsp([1, 2, 3]))

cities = [1, 2, 3]
graph = {
    1: {2: 10, 3: 15},
    2: {1: 10, 3: 20},
    3: {1: 15, 2: 20}
}
result = brute_force_tsp(cities, graph)
print("Optimal TSP Path:", result['path'])
print("Total Distance:", result['distance'])

print("is_subset_sum([1, 2, 3]): ", is_subset_sum([1, 2, 3], 3, 9))
input_set = [3, 34, 4, 12, 5, 2]
target_sum = 9
is_subset_possible = is_subset_sum(input_set, len(input_set), target_sum)
print(f"Subset with sum {target_sum} exists: {is_subset_possible}")
